chasm/tree.go

```diff
@@ -137,6 +152,378 @@ func NewEmptyTree(
+				return err
+			}
+
+			childrenToKeep[fieldN] = struct{}{}
+		case chasmCollectionTypePrefix:
+			childrenToKeep[fieldN] = struct{}{}
+			// TODO: need to go over every item in collection and update children for it.
+			panic("not implemented")
+		}
+	}
+
+	err := n.deleteChildren(removedPaths, childrenToKeep, nodePath)
+	return err
+}
+
+func (n *Node) deleteChildren(removedPaths *[]string, childrenToKeep map[string]struct{}, currentPath []string) error {
+	for childName, childNode := range n.children {
+		if _, childToKeep := childrenToKeep[childName]; !childToKeep {
+			if err := childNode.deleteChildren(removedPaths, nil, append(currentPath, childName)); err != nil {
+				return err
+			}
+			path, err := n.pathEncoder.Encode(n, append(currentPath, childName))
+			if err != nil {
+				return err
+			}
+			*removedPaths = append(*removedPaths, path)
+			// If parent is about to be removed, it must not have any children.
+			// TODO: softassert: len(childNode.children)==0
+			delete(n.children, childName)
+		}
+	}
+	return nil
+}
+
+func (n *Node) serializeDataNode() error {
+	protoValue, ok := n.value.(proto.Message)
+	if !ok {
+		return serviceerror.NewInternal("only support proto.Message as chasm data")
+	}
+
+	blob, err := serialization.ProtoEncodeBlob(protoValue, enumspb.ENCODING_TYPE_PROTO3)
+	if err != nil {
+		return err
+	}
+
+	n.serializedValue.GetDataAttributes().Data = blob
+	return nil
+}
+
+// deserialize initializes the node's value from its serializedValue.
+// If value is of component type, it initializes every chasm.Field of it and sets node field but not value field.
+// deserialize is not recursive; i.e., must be called on every node separately.
+func (n *Node) deserialize(
+	valueT reflect.Type,
+) error {
+	if err := validateType(valueT); err != nil {
+		return err
+	}
+
+	switch n.serializedValue.Attributes.(type) {
+	case *persistencespb.ChasmNode_ComponentAttributes:
+		return n.deserializeComponentNode(valueT)
+	case *persistencespb.ChasmNode_DataAttributes:
+		return n.deserializeDataNode(valueT)
+	case *persistencespb.ChasmNode_CollectionAttributes:
+		panic("not implemented")
+	case *persistencespb.ChasmNode_PointerAttributes:
+		// TODO: return serviceerror.NewInternal(...) instead.
+	}
+	return nil
+}
+
+func (n *Node) deserializeComponentNode(
+	valueT reflect.Type,
+) error {
+	valueV := reflect.New(valueT.Elem())
+	protoMessageFound := false
+	for i := 0; i < valueT.Elem().NumField(); i++ {
+		fieldV := valueV.Elem().Field(i)
+		fieldT := fieldV.Type()
+		if fieldT.AssignableTo(protoMessageT) {
+			if protoMessageFound {
+				return serviceerror.NewInternal("only one proto field allowed in component")
+			}
+			protoMessageFound = true
+
+			value, err := n.unmarshalProto(n.serializedValue.GetComponentAttributes().GetData(), fieldT)
+			if err != nil {
+				return err
+			}
+			fieldV.Set(value)
+			continue
+		}
+
+		// chasm.Field field must be a pointer, i.e. *chasm.Field[T].
```

Comment:
I don't think it has to be a pointer. As long as the registered component that's using chasm.Field is a pointer we should be good I think. 
Is there a case you have in mind that doesn't work when Field is used without a pointer?

Reply:
I just don't want to support both cases and existing `Field` constructors return `*Field`. I agree that `Field` look more natural in component definition than `*Field`. I am flipping it to use `Field` and forbid `*Field`.